eth/signer.go

```diff
@@ -149,7 +149,35 @@ func (s *Signer) SignWithNonce(buf []byte) ([]byte, [32]byte, error) {
 	return sig, nonce, nil
 }
 
-// SignAddress is a convenience wrapper for s.CompactSign(addr.Bytes()).
-func (s *Signer) SignAddress(addr common.Address) ([]byte, error) {
-	return s.CompactSign(addr.Bytes())
+// toEthSignedMessageHash converts a given message to conform to the signed data
+// standard according to EIP-191.
+func toEthPersonalSignedMessage(message []byte) []byte {
+	prefix := []byte(fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(message)))
+	return append(prefix, message...)
+}
+
+// PersonalSign returns an EIP-191 conform personal ECDSA signature of buf
+// Convenience wrapper for s.CompactSign(toEthPersonalSignedMessage(buf))
+func (s *Signer) PersonalSign(buf []byte) ([]byte, error) {
+	return s.CompactSign(toEthPersonalSignedMessage(buf))
+}
+
+// PersonalSignWithNonce generates a 32-byte nonce with crypto/rand and returns
+// s.PersonalSign(append(buf, nonce)).
+func (s *Signer) PersonalSignWithNonce(buf []byte) ([]byte, [32]byte, error) {
+	var nonce [32]byte
+	if n, err := rand.Read(nonce[:]); n != 32 || err != nil {
+		return nil, nonce, fmt.Errorf("read 32 random bytes: got %d bytes with err %v", n, err)
+	}
```

Comment:
I think the abstractions can be improved a bit to aid in understanding / reading the code (I hark on about this a lot because I'm a strong proponent of lower cognitive load leading to fewer bugs).

It feels like we're ending up with lots of different versions of signing, which suggests that an options struct might be a cleaner approach. We can still expose convenience functions, but centralise the implementation (in fact I think it's best to not export the options without more thorough thought around API / default behaviour). Something like this:

```Go
type signOpts struct {
  raw, compact, personal, withNonce bool
}

func (s *Signer) sign(buf []byte, opts signOpts) ([]byte, *[32]byte, error) {
  if opts.personal {
    // add prefix
  }

  if !opts.raw {
    buf = crypto.Keccak256(buf)
  }

  var nonce *[32]byte
  if opts.withNonce {
    nonce = new([32]byte)
    …
  }

  // sign

  if !opts.compact {
    return sig, nonce, nil
  }

  // logic from compact
}
```

I know that the `*[32]byte` pointer is weird, but it's the only way to return a nil nonce if it's not requested, other than creating separate functions.

Reply:
I agree. This is much easier to reason than having different implementations.